### DISCUSSION TITLE
Custom Hooks

### DISCUSSION THREAD

**Discussion by takinola** [pocketbase/pocketbase#3353] [Q&A] [answered]
Is it possible to trigger a custom hook on an event not related to a record being created or modified?  For example, I want to be able to send an email to a user when a particular event happens in the app.  
```
//  /pb_hooks/main.pb.js

myCustomHook((e)=>{
  // do stuff to send email
})

-----------------------------
// /app.js

// invoke myCustomHook
const message = {to: ...., message: ...}
myCustomHook(message);
```

The closest I can see if the [Console Commands](https://pocketbase.io/docs/js-console-commands/) but those seem only accessible from the command line and not from an app instance

**Comment by ganigeorgiev** [comment]
I'm not entirely sure that I understand what you are trying to do, but there is no option to register a custom hook with the JS `pb_hooks`. The idea of the hooks is to provide some extendability when often direct access and modification to the code internals is not possible, but I don't think this is the case here as you have full control over you own JS hooks code.

If you just want to call a common function you can put it in a local JS module and load it with `require` whenever you want to access it inside your handlers. See the example in https://pocketbase.io/docs/js-overview/#loading-modules.

```js
// pb_hooks/utils.js
module.exports = {
    // if you want access to the $app instance you can pass it as an argument
    hello: (name) => {
        console.log("Hello " + name)
    }
}

// pb_hooks/main.pb.js
onAfterBootstrap((e) => {
    const utils = require(`${__hooks}/utils.js`)

    utils.hello("world")
})
```

**Reply by ganigeorgiev** [reply]
_Additionally if you haven't already you can check https://pocketbase.io/docs/js-sending-emails/._

**Reply by takinola** [reply]
My scenario is that I am trying to implement a user invite feature.  An admin can send out an invitation to an email address so they can claim an account.  To implement this, the process is as follows

1. Create a user account (with a dummy password)
2. Send a verification request to the user email
3. When the user clicks on the verification request, they are directed to change the password of the account (since the user does not have access to the dummy password)
4. When the password reset token has been verified and the user password has been set, mark the user account as verified and log them in

In this process, I am currently repurposing the password reset process with a custom email template.  However, this makes it hard to implement a normal password reset process since the email template has been modified for this other purpose.  Instead, I would prefer to create a custom hook event designed specifically for the user invite use case.

**Reply by ganigeorgiev** [reply]
I'm still not sure that I understand what do you mean by _"I would prefer to create a custom hook event"_.
Even if you create a custom hook, where and what will trigger this hook? I'm just failing to understand how a custom hook will work for your use case and why a simple function call will not suffice.
Could you provide a pseudo-code with the desired behavior?

**Reply by ganigeorgiev** [reply]
If you want to create a custom route (if that is what you mean by "custom hook"), you can check https://pocketbase.io/docs/js-routing/#registering-new-routes.

For sending requests to custom routes with the SDKs, see https://pocketbase.io/docs/js-routing/#sending-request-to-custom-routes-using-the-sdks.

**Reply by ganigeorgiev** [reply, marked as answer]
If you want to use the password reset action to send a custom invite email, you can do something like (_I haven't tested it and there are other possible approaches, eg. using the `onMailerBeforeRecordResetPasswordSend` hook with a conditional mail message based on some user record flag_):

```js
onRecordBeforeRequestPasswordResetRequest((e) => {
    const isInvite = !!e.httpContext.queryParam("invite");
    if (!isInvite || e.record.verified()) {
        return; // proceed with the regular hook execution
    }

    // send your custom email...
    const message = new MailerMessage({
        from: {
            address: $app.settings().meta.senderAddress,
            name:    $app.settings().meta.senderName,
        },
        to:      [{address: e.record.email()}],
        subject: "YOUR_SUBJECT...",
        html:    "YOUR_HTML_BODY...",
    })

    $app.newMailClient().send(message)

    e.httpContext.noContent(204);

    return false // stops the execution chain
}, "users")
```

On the client side to trigger the above you'll need to pass a `invite` query parameter:

```js
await pb.collection("users").requestPasswordReset("test@example.com", { invite: 1 })
```

**Reply by takinola** [reply]
The approach in the last code block is what I am looking for.  I was not aware you could include query parameters in the [requestPasswordReset](https://pocketbase.io/docs/api-records/#request-password-reset) function.  Is that documented anywhere?

Thanks for the help

**Reply by takinola** [reply]
Is it possible to retrieve the password reset token from [onRecordBeforeRequestPasswordResetRequest](https://pocketbase.io/jsvm/interfaces/core.RecordRequestPasswordResetEvent.html)?

The event passed into this hook does not have the e.meta.token property (like the [onMailerBeforeRecordResetPasswordSend](https://pocketbase.io/jsvm/functions/onMailerBeforeRecordResetPasswordSend.html) hook)

Without the token, it does not seem possible to recreate a password reset email in this hook.

**Reply by ganigeorgiev** [reply]
> Is it possible to retrieve the password reset token from [onRecordBeforeRequestPasswordResetRequest](https://pocketbase.io/jsvm/interfaces/core.RecordRequestPasswordResetEvent.html)?

I forgot to mention that, sorry. The token is not generated yet at this stage of the hook execution chain (_it is created right before sending the email_).

You can create the token manually by calling [`$tokens.recordResetPasswordToken($app, e.record)`](https://pocketbase.io/jsvm/functions/_tokens.recordResetPasswordToken.html).

_Also small side-note about the above code sample - since it is a request hook in your specific case you'll need also to wrote a response body, aka. `e.httpContext.noContent(204)` (I've updated the example)._

**Reply by takinola** [reply]
Works perfectly.  You are a lifesaver!

What is the purpose of this line?
````
e.httpContext.noContent(204);
````

I'm not clear on why (if) it is required

**Reply by ganigeorgiev** [reply]
It ensures that the request will have a proper HTTP response returned to the client since the default finalizer for this case is skipped.
